### PR TITLE
Feature: add XML sitemaps for SEO

### PR DESCRIPTION
**Pull request by oubiwann**
This feature adds sitemaps to CMR search, building on the new HTML pages feature which provided directories of collection landing pages (see #32).

In particular, this change adds the following:

* a `site/sitemap.xml` - includes locations for static URLs (in particular, provides links to directories of landing page links, but also includes docs)
* a `site/collections/directory/:provider-id/:tag/sitemap.xml` - includes locations for supported provider+tag combinations
* a `sitemap.xml` - this is a "master" sitemap that uses the sitemap index XML schema to combine a collection of sitemaps; this is what should be submitted to Google

Note that this feature adds a new Clojure dependency upon an XML XSD validator, so it will be necessary to run `lein modules do clean, install, clean`.

### CODE REVIEW

File: search-app/docs/site.md [hunk 1]
```diff
@@ -14,12 +14,15 @@ See also the [API Documentation](search_api_docs.html).
```
Comment (chris-durbin) [inline search-app/docs/site.md]: I noticed some typos on https://cmr.sit.earthdata.nasa.gov/search/site/docs 

1) Two misspellings - Doucmentation
2) Change sentence to "How to access the search API via an HTTP client"

File: search-app/resources/templates/sitemap-provider-tag.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ {% for link in links %}
+  <url>
+    <loc>{{ link.href }}</loc>
+    <lastmod>{{ datestamp }}</lastmod>
```
Comment (chris-durbin) [inline search-app/resources/templates/sitemap-provider-tag.xml]: It seems like we should include the <changefreq> tag as well. Are we regenerating the sitemap pages daily?

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 3]
```diff
@@ -145,97 +159,94 @@
...
+        response (get-response url-path)
+        body (:body response)]
+    (testing "check the status and links for PROV3"
       (is (= 200 (:status response)))
-      (is (string/includes? body expected-provider1-level-links))
-      ;; The collections not tagged with eosdis shouldn't show up
-      (is (not (string/includes? body notexpected-provider-level-link)))
-      ;; This page should also have a header link
+      (is (string/includes? body expected-provider1-level-links)))
+    (testing "the collections not taged with eosdis shouldn't show up"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: tagged typo

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 4]
```diff
@@ -145,97 +159,94 @@
...
+        response (get-response url-path)
+        body (:body response)]
+    (testing "check the status and links for PROV3"
       (is (= 200 (:status response)))
-      (is (string/includes? body expected-provider3-level-links))
-      ;; The collections not tagged with eosdis shouldn't show up
-      (is (not (string/includes? body notexpected-provider-level-link)))
-      ;; This page should also have a header link
+      (is (string/includes? body expected-provider3-level-links)))
+    (testing "the collections not taged with eosdis shouldn't show up"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: tagged typo

File: system-int-test/test/cmr/system_int_test/search/sitemaps.clj [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            [cmr.system-int-test.system :as s]
+            [cmr.transmit.config :as transmit-config]
+            [cmr.umm-spec.models.umm-common-models :as cm]
+            [cmr.umm-spec.test.expected-conversion :as exp-conv]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants and general utility functions for the tests
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(def ^{:doc "We don't call to `(transmit-config/application-public-root-url)`
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/sitemaps.clj]: Our coding style is:
```
(def ^:private base-url
  "We don't call to ..."
  (format "...)
```

File: system-int-test/test/cmr/system_int_test/search/sitemaps.clj [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        url-path (format
+                  "site/collections/directory/%s/%s/sitemap.xml"
+                  provider tag)
+        response (get-response url-path)
+        body (:body response)]
+    (testing "presence and content of sitemap.xml file"
+      (is (= 200 (:status response)))
+      (is (not (string/includes? body "http://dx.doi.org/doi5</loc>")))
+      (is (not (string/includes? body "http://dx.doi.org/doi5</loc>"))))
+    (testing "the collections not taged with eosdis shouldn't show up"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/sitemaps.clj]: Global replace taged with tagged

File: search-app/src/cmr/search/site/routes.clj [hunk 7]
```diff
@@ -23,6 +23,12 @@
         (GET "/"
              {context :request-context}
              (pages/home context))
+        (GET "/sitemap.xml"
+             {context :request-context}
+             (pages/sitemap-master context))
+        (GET "/site/sitemap.xml"
```
Comment (chris-durbin) [inline search-app/src/cmr/search/site/routes.clj]: Why do we have this top level XML file in addition to the sitemap master XML file?

File: search-app/src/cmr/search/site/routes.clj [hunk 8]
```diff
@@ -23,6 +23,12 @@
         (GET "/"
              {context :request-context}
              (pages/home context))
+        (GET "/sitemap.xml"
```
Comment (chris-durbin) [inline search-app/src/cmr/search/site/routes.clj]: We should pay attention to how long it takes to request each of these routes. For example for a provider with more than a thousand collections, how long does it take to generate this XML?

I'm thinking if the time is excessive we might want some kind of background job that periodically generates the XML behind the scenes and we just serve it up from cache.

Comment (oubiwann) [inline search-app/src/cmr/search/site/routes.clj]: That's a good idea.

File: search-app/src/cmr/search/site/routes.clj (hunk 7 above)
Comment (oubiwann) [inline search-app/src/cmr/search/site/routes.clj]: `sitemap.xml` is the master -- it uses the sitemap "index" schema (as opposed to the regular sitemap schema). It is my understanding that the sitemap index schema only supports collections of sitemaps, not collections of content locations.

File: system-int-test/test/cmr/system_int_test/search/sitemaps.clj (hunk 6 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/sitemaps.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/search/sitemaps.clj (hunk 5 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/sitemaps.clj]: Updated.

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 4 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Fixed.

File: search-app/resources/templates/sitemap-provider-tag.xml (hunk 2 above)
Comment (chris-durbin) [inline search-app/resources/templates/sitemap-provider-tag.xml]: I missed the tag in that comment. 😄 

<changefreq>

But after reading through everything it looks like we're regenerating on every request which would imply "always". I don't think we'd want anything to crawl more than once a day though, so maybe we should just specify "daily". Do you think it makes sense including and would "daily" be ok?

Comment (oubiwann) [inline search-app/resources/templates/sitemap-provider-tag.xml]: I think that is perfectly reasonable.

Comment (oubiwann) [inline search-app/resources/templates/sitemap-provider-tag.xml]: Added.

File: search-app/docs/site.md (hunk 1 above)
Comment (oubiwann) [inline search-app/docs/site.md]: Fixed.

Also did a visual check on the other docs with spell-check enabled and found some old typos -- those are also fixed now.

**Comment by oubiwann** [comment]
Thanks! Issues addressed.

Will re-run the tests and then merge if all okay. The demo uses changes made here, so I'll submit a PR for the feature demo once this code is in master.

**Pull request merged**